glmmTMB/src/distrib.h

```diff
@@ -0,0 +1,265 @@
+                             // ATOMIC_DOUBLE
+                             ty[0] = Rf_logspace_sub(exp(tx[0]), 0.);
+                             ,
+                             // ATOMIC_REVERSE
+                             px[0] = exp( logspace_add(tx[0], tx[0]-ty[0]) ) * py[0];
+                             )
+  template<class Type>
+  Type logit_invcloglog(Type x) {
+    CppAD::vector<Type> tx(1);
+    tx[0] = x;
+    return logit_invcloglog(tx)[0];
+  }
+
+  /* y(x) = logit_pnorm(x) := logit( pnorm(x) ) =
+     pnorm(x, lower.tail=TRUE,  log.p=TRUE) -
+     pnorm(x, lower.tail=FALSE, log.p=TRUE)
+
+     y'(x) = dnorm(x) * ( (1+exp(y)) + (1+exp(-y)) )
+
+  */
+  double logit_pnorm(double x) {
+    double log_p_lower, log_p_upper;
+    Rf_pnorm_both(x, &log_p_lower, &log_p_upper, 2 /* both tails */, 1 /* log_p */);
+    return log_p_lower - log_p_upper;
+  }
+  TMB_ATOMIC_VECTOR_FUNCTION(
+                             // ATOMIC_NAME
+                             logit_pnorm
+                             ,
+                             // OUTPUT_DIM
+                             1,
+                             // ATOMIC_DOUBLE
+                             ty[0] = logit_pnorm(tx[0])
+                             ,
+                             // ATOMIC_REVERSE
+                             Type zero = 0;
+                             Type tmp1 = logspace_add(zero, ty[0]);
+                             Type tmp2 = logspace_add(zero, -ty[0]);
+                             Type tmp3 = logspace_add(tmp1, tmp2);
+                             Type tmp4 = dnorm(tx[0], Type(0), Type(1), true) + tmp3;
+                             px[0] = exp( tmp4 ) * py[0];
+                             )
+  template<class Type>
+  Type logit_pnorm(Type x) {
+    CppAD::vector<Type> tx(1);
+    tx[0] = x;
+    return logit_pnorm(tx)[0];
+  }
+
+  /* Calculate variance in compois family using
+
+     V(X) = (logZ)''(loglambda)
+
+  */
+  double compois_calc_var(double mean, double nu){
+    using atomic::compois_utils::calc_loglambda;
+    using atomic::compois_utils::calc_logZ;
+    double loglambda = calc_loglambda(log(mean), nu);
+    typedef atomic::tiny_ad::variable<2, 1, double> ADdouble;
+    ADdouble loglambda_ (loglambda, 0);
+    ADdouble ans = calc_logZ<ADdouble>(loglambda_, nu);
+    return ans.getDeriv()[0];
+  }
+
+  /* Simulate from zero-truncated Conway-Maxwell-Poisson distribution */
+  template<class Type>
+  Type rtruncated_compois2(Type mean, Type nu) {
+    int nloop = 10000;
+    int counter = 0;
+    Type ans = rcompois2(mean, nu);
+    while(ans < 1. && counter < nloop) {
+      ans = rcompois2(mean, nu);
+      counter++;
+    }
+    if(ans < 1.) warning("Zeros in simulation of zero-truncated data. Possibly due to low estimated mean.");
+    return ans;
+  }
+
+  /* Simulate from tweedie distribution */
+  template<class Type>
+  Type rtweedie(Type mu, Type phi, Type p) {
+    // Copied from R function tweedie::rtweedie
```

Comment:
Maybe we should have added Peter Dunn as a contributor since the Tweedie simulation code was copied from a function he wrote. Should I email him to ask?